File: plugin/src/sync/crypto.ts
/**
 * Client-side vault encryption: Argon2id key derivation + AES-256-GCM.
 *
 * Encryption flow:
 * 1. Derive key: Argon2id(passphrase, vault_salt) → 256-bit key
 * 2. Encrypt: AES-256-GCM(key, random_nonce, plaintext) → nonce + ciphertext + tag
 * 3. Upload ciphertext to server (server never sees plaintext)
 *
 * Implementation note: Argon2 is imported at runtime via argon2-browser.
 * For the MVP, we use PBKDF2 as a fallback (available via SubtleCrypto).
 */

const NONCE_LENGTH = 12; // AES-GCM nonce
const KEY_LENGTH = 32; // 256 bits
const PBKDF2_ITERATIONS = 600000;

/**
 * Derive an AES-256 key from a passphrase and salt using PBKDF2.
 * (Argon2id will be added when argon2-browser is integrated)
 */
export async function deriveKey(
  passphrase: string,
  salt: string
): Promise<CryptoKey> {
  const enc = new TextEncoder();
  const keyMaterial = await crypto.subtle.importKey(
    "raw",
    enc.encode(passphrase),
    "PBKDF2",
    false,
    ["deriveKey"]
  );
  return crypto.subtle.deriveKey(
    {
      name: "PBKDF2",
      salt: enc.encode(salt),
      iterations: PBKDF2_ITERATIONS,
      hash: "SHA-256",
    },
    keyMaterial,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"]
  );
}

/**
 * Encrypt data with AES-256-GCM. Returns nonce + ciphertext.
 */
export async function encrypt(
  key: CryptoKey,
  plaintext: ArrayBuffer
): Promise<ArrayBuffer> {
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LENGTH));
  const ciphertext = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    plaintext
  );
  // Prepend nonce to ciphertext
  const result = new Uint8Array(NONCE_LENGTH + ciphertext.byteLength);
  result.set(nonce, 0);
  result.set(new Uint8Array(ciphertext), NONCE_LENGTH);
  return result.buffer;
}

/**
 * Decrypt data with AES-256-GCM. Input is nonce + ciphertext.
 */
export async function decrypt(
  key: CryptoKey,
  data: ArrayBuffer
): Promise<ArrayBuffer> {
  const bytes = new Uint8Array(data);
  const nonce = bytes.slice(0, NONCE_LENGTH);
  const ciphertext = bytes.slice(NONCE_LENGTH);
  return crypto.subtle.decrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    ciphertext
  );
}

/**
 * Generate a random salt for vault encryption.
 */
export function generateSalt(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(32));
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

/**
 * Hash a derived key for server-side passphrase verification.
 * The server stores this hash — it can verify the passphrase is correct
 * without ever seeing the actual encryption key.
 */
export async function hashKey(key: CryptoKey): Promise<string> {
  const exported = await crypto.subtle.exportKey("raw", key);
  // Re-import as non-extractable would fail, so hash the raw bytes
  const hash = await crypto.subtle.digest("SHA-256", exported);
  const hashArray = Array.from(new Uint8Array(hash));
  return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
}

File: cli/src/obsidian_sync_cli/sync/crypto.py
"""Encryption utilities for vault sync.

Uses PBKDF2-SHA256 for key derivation and AES-256-GCM for encryption.
Mirrors the server and plugin crypto implementations for cross-client
compatibility.
"""

import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

_KDF_ITERATIONS = 600_000
_KEY_LENGTH = 32  # 256 bits
_NONCE_LENGTH = 12  # 96 bits, standard for AES-GCM


def derive_key(passphrase: str, salt: str) -> bytes:
    """Derive a 32-byte encryption key from a passphrase and hex salt.

    Uses PBKDF2-SHA256 with 600 000 iterations.
    """
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=_KEY_LENGTH,
        salt=bytes.fromhex(salt),
        iterations=_KDF_ITERATIONS,
    )
    return kdf.derive(passphrase.encode("utf-8"))


def encrypt(key: bytes, plaintext: bytes) -> bytes:
    """Encrypt *plaintext* with AES-256-GCM.

    Returns ``nonce || ciphertext`` (12 + len(plaintext) + 16 bytes).
    """
    nonce = os.urandom(_NONCE_LENGTH)
    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(nonce, plaintext, None)
    return nonce + ciphertext


def decrypt(key: bytes, data: bytes) -> bytes:
    """Decrypt data produced by :func:`encrypt`.

    Expects the first 12 bytes to be the nonce followed by the GCM
    ciphertext (including the 16-byte auth tag).
    """
    nonce = data[:_NONCE_LENGTH]
    ciphertext = data[_NONCE_LENGTH:]
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(nonce, ciphertext, None)


def generate_salt() -> str:
    """Return 32 random bytes encoded as a hex string."""
    return os.urandom(32).hex()


def hash_key(key: bytes) -> str:
    """Return the SHA-256 hex digest of *key*.

    Used for server-side passphrase verification without exposing the
    actual encryption key.
    """
    return hashlib.sha256(key).hexdigest()
